### PR TITLE
update encrypt message use d-chat

### PR DESCRIPTION
**Pull request by iheron**
use this can send/receive message with d-chat

### CODE REVIEW

**Comment by RealJohnSmith** [comment]
I don't understand what is this issue trying to solve. Please provide more information.
The encryption is already there, in place.

Try to be more informative with your text. For example "update encrypt message use d-chat" doesn't tell me absolutely anything. Why would you want to update that? What is the difference between general encryption and some d-chat one? Why should that be solved in a Sdk and not d-chat itself?

File: src/main/java/jsmith/nknsdk/network/ClientTunnel.java [hunk 1]
```diff
@@ -249,18 +249,19 @@ private boolean setupWsConnection() {
...
-                            cm.onInboundMessage(from, payload);
-                            break;
-
-                        default:
-                            LOG.warn("Got invalid payload type {}, ignoring", payload.getType());
-                            break;
-                    }
+                    cm.onInboundMessage(from, pldMsg);
+
+//                    switch (payload.getType()) {
```
Comment (yilunzhang) [inline src/main/java/jsmith/nknsdk/network/ClientTunnel.java]: We removing this part of logic?

**Comment by yilunzhang** [comment]
Let to try to help a bit.
Please correct me if I'm wrong but based on my understanding, this PR seems to have nothing to do with d-chat but is trying to fix 2 problems in the java sdk:
1. The order of EncryptedMessage and Payload are reversed, i.e. Payload should be included in EncryptedMessage but not the opposite. When a message is encrypted, everything in Payload should be encrypted, include type, pid, reply_to_pid, etc, otherwise it's leaking some information about the message. For example, a middle man will know if a message is a reply or not, and if not, he will know the message pid (even if he cannot understand message content) and thus can send random-reply with the correct reply_to_pid. It's always better to leak as less info as possible for encrypted message.
2. It seems that you are calling the wrong TweetNaclFast API which caused the msg to be not compatible with standard TweetNacl protocol.

After fixing these 2 issues the java sdk should be compatible with other nkn client sdk now.

And I agree with @RealJohnSmith  that we should put these information in the PR description.

**Comment by RealJohnSmith** [comment]
1. is a good point, I think I was copying this functionality from js client. I will have a look at that.
2. I think  tested the my SDK agains the JS one, to make sure the encryption and stuff works correctly. Maybe not. 
I will try to find some time this week to check it out

File: src/main/java/jsmith/nknsdk/network/ClientMessages.java [hunk 2]
```diff
@@ -9,6 +9,7 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.awt.SystemTray;
```
Comment (RealJohnSmith) [inline src/main/java/jsmith/nknsdk/network/ClientMessages.java]: What and why is this?

File: src/main/java/jsmith/nknsdk/network/ClientMessages.java [hunk 3]
```diff
@@ -161,28 +162,29 @@ public void onMessage(Function<NKNClient.ReceivedMessage, Object> listener) {
...
 
-        MessagesP.EncryptedMessage encryptedMessage;
+        ByteString msg;
         try {
-            encryptedMessage = MessagesP.EncryptedMessage.parseFrom(message.getData());
-        } catch (InvalidProtocolBufferException e) {
-            LOG.warn("Received invalid message, ignoring");
+            msg = ClientEnc.decryptMessage(from, pldMsg, ct.identity.wallet);
+        } catch (NKNClientException e) {
+            System.out.println("Failed to decrypt message, dropping");
```
Comment (RealJohnSmith) [inline src/main/java/jsmith/nknsdk/network/ClientMessages.java]: Don't ever use random system prints in code. That's what proper logs are for

File: src/main/java/jsmith/nknsdk/network/ClientMessages.java [hunk 4]
```diff
@@ -161,28 +162,29 @@ public void onMessage(Function<NKNClient.ReceivedMessage, Object> listener) {
...
-
-        ByteString data;
+        final MessagesP.Payload message;
         try {
-            data = ClientEnc.decryptMessage(from, encryptedMessage, ct.identity.wallet);
-        } catch (NKNClientException e) {
-            LOG.warn("Failed to decrypt message, dropping");
+            message = MessagesP.Payload.parseFrom(msg);
+        } catch (InvalidProtocolBufferException e) {
+            System.out.println("Payload.parseFrom invalid message, ignoring");
```
Comment (RealJohnSmith) [inline src/main/java/jsmith/nknsdk/network/ClientMessages.java]: ^ Same

File: src/main/java/jsmith/nknsdk/network/ClientMessages.java [hunk 5]
```diff
@@ -282,28 +284,42 @@ public void onInboundMessage(String from, MessagesP.Payload message) {
...
                         .setPid(messageID)
                         .setReplyToPid(replyToMessageID)
-                        .setData(ClientEnc.encryptMessage(Collections.singletonList(d), message, ct.identity.wallet, NKNClient.EncryptionLevel.ENCRYPT_ONLY_UNICAST))
+                        .setData(message)
                         .setNoAck(noAck)
                         .build();
+                final MessagesP.EncryptedMessage encryptedMessage;
+                try {
+                    encryptedMessage = MessagesP.EncryptedMessage.parseFrom(ClientEnc.encryptMessage(Collections.singletonList(d), payload.toByteString(), ct.identity.wallet, NKNClient.EncryptionLevel.ENCRYPT_ONLY_UNICAST));
+                    System.out.println("encryptedMessage.toByteString():" + encryptedMessage.toByteString());
```
Comment (RealJohnSmith) [inline src/main/java/jsmith/nknsdk/network/ClientMessages.java]: ^

File: src/main/java/jsmith/nknsdk/network/ClientMessages.java [hunk 6]
```diff
@@ -282,28 +284,42 @@ public void onInboundMessage(String from, MessagesP.Payload message) {
...
+                    promises.addAll(sendOutboundMessage(Collections.singletonList(d), messageID, encryptedMessage.toByteString()));
+                } catch (InvalidProtocolBufferException e) {
+                    e.printStackTrace();
```
Comment (RealJohnSmith) [inline src/main/java/jsmith/nknsdk/network/ClientMessages.java]: Again, no prints, logs.

File: src/main/java/jsmith/nknsdk/network/ClientTunnel.java (hunk 1 above)
Comment (RealJohnSmith) [reply src/main/java/jsmith/nknsdk/network/ClientTunnel.java]: Hm, thats fair. Before encryption there was different handling of ACK, iirc.
However, it is better to just delete it, don't comment out bits of code (don't keep dead code in the codebase. We have git if we ever need to see older version)

**Review by RealJohnSmith — CHANGES_REQUESTED** [review]
In general, it looks ok. I will do more testing hopefully in a few days, however, I have a few comments which I am not willing to merge in (logs and stuff). Either change it and make new PR (rebase commit, so it is not in history), or I will re-do the code myself and merge that. (using yours as inspiration, of course)

* There might be more requested changes in the future when I actually try to run this.

**Comment by iheron** [comment]
Sorry. My java is not very skilled, trouble you review code.😀

**Comment by RealJohnSmith** [comment]
Hi.
Your code didn't work (didn't pass even the SimpleExample). I tried to make it work for quite some time and I failed, so I rewrote the entire thing myself. Please have a look if it solves your problem and let me know. It is commit 9a9303ab4ab78981e38a7a16a0ef1dea63efbe16 (latest on master at the time of writing this comment)

**Comment by iheron** [comment]
I think your code is better then me. But decryptMessage and encryptMessage function must be modified. Or incompatible with other nkn client sdk. 
```
public static ByteString decryptMessage(String from, MessagesP.EncryptedMessage enc, Wallet wallet) throws NKNClientException {
        if (enc.getEncrypted()) {

            final byte[] key = wallet.getSharedKey(from);
            final byte[] msg = enc.getPayload().toByteArray();
            final byte[] nonce = enc.getNonce().toByteArray();

            byte[] bytes = new TweetNaclFast.SecretBox(key).open(msg, nonce);
            return ByteString.copyFrom(bytes, 0, bytes.length);

        } else {
            return enc.getPayload();
        }
    }
```
If want test code, you can install D-Chat(https://chrome.google.com/webstore/detail/d-chat/glnmkakdjcognfgonjfcklpmjiobijii). Use D-Chat send a private message to java sdk. Receive message like `Client A: New encrypted text from heron.25ac590eaca614a0ba4c4387d8514a0b54e948d120c6ff49564e7830c9dec929
  ==> {"contentType":"text","id":"4a7d9801-a8ce-4a2f-901d-586f118e3401","timestamp":"Thu, 28 Nov 2019 11:03:02 GMT","content":"hello world","isPrivate":true}`,

**Comment by RealJohnSmith** [comment]
It should work now.

**Comment by iheron** [comment]
Yeah, now can decrypt the message from other nkn client.👍

**Pull request closed without merging**